### DISCUSSION TITLE
Can I use a custom validation library?

### DISCUSSION THREAD

**Discussion by brielov** [trpc/trpc#2199] [Q&A] [answered]
Hi, I have my own validation library which is heavily inspired by zod and superstruct. It is both much smaller and much faster. The API is also very similar. Can I use it or am I limited to just zod / yup / superstruct?

This is my [library](https://github.com/brielov/typed) in case you're wondering.

**Comment by KATT** [comment, marked as answer]
Should be possible, we also allow for generic callback fns, so you could probably make an adapter there. You can also make an adapter in tRPC.

If you do, we'll only accept changes against the `next`-branch for these things right now.

It should be fairly straight-forward to add your own adapter:

- https://github.com/trpc/trpc/blob/next/packages/server/src/core/parser.ts
- https://github.com/trpc/trpc/blob/next/packages/server/src/core/internals/getParseFn.ts
